src/v/cluster/log_eviction_stm.cc

```diff
@@ -40,82 +41,124 @@ log_eviction_stm::log_eviction_stm(
   ss::abort_source& as,
   storage::kvstore& kvstore)
   : persisted_stm("log_eviction_stm.snapshot", logger, raft, kvstore)
-  , _logger(logger)
-  , _as(as) {}
+  , _as(as)
+  , _queue(max_event_queue_size) {}
 
 ss::future<> log_eviction_stm::start() {
     ssx::spawn_with_gate(_gate, [this] { return monitor_log_eviction(); });
     ssx::spawn_with_gate(
-      _gate, [this] { return write_raft_snapshots_in_background(); });
+      _gate, [this] { return handle_log_eviction_events(); });
     return persisted_stm::start();
 }
 
 ss::future<> log_eviction_stm::stop() {
-    _reap_condition.broken();
+    _queue_mutex.broken();
+    _queue.abort(std::make_exception_ptr(ss::abort_requested_exception()));
     co_await persisted_stm::stop();
 }
 
-ss::future<> log_eviction_stm::write_raft_snapshots_in_background() {
+ss::future<> log_eviction_stm::enqueue_eviction_event(
+  model::offset offset, retry_event_until_success wait) {
+    vlog(
+      _log.trace,
+      "log eviction event at offset: {}, waiting for truncation: {}",
+      offset,
+      wait);
+
+    auto u = co_await _queue_mutex.get_units();
+    co_await _queue.push_eventually(eviction_event{
+      .prefix_truncate_offset = offset, .wait_for_success = wait});
+}
+
+ss::future<> log_eviction_stm::handle_log_eviction_events() {
+    static constexpr auto retry_backoff_time = 5s;
     /// This method is executed as a background fiber and it attempts to write
     /// snapshots as close to effective_start_offset as possible.
     auto gh = _gate.hold();
-    bool wait_with_timeout = false;
+
     while (!_as.abort_requested()) {
         /// This background fiber can be woken-up via apply() when special
         /// batches are processed or by the storage layer when local
         /// eviction is triggered.
         try {
-            static const auto eviction_event_wait_time = 5s;
-            if (wait_with_timeout) {
-                /// If after call to \ref do_write_snapshot() has not
-                /// truncated up until the desired point, it will be OK to
-                /// iterate again after some time to retry. Maybe
-                /// max_collectible_offset has moved forward.
-                co_await _reap_condition.wait(eviction_event_wait_time);
-            } else {
-                /// Last truncation had completed with success, there is no need
-                /// to wait for any other condition then for notify() to be
-                /// called.
-                co_await _reap_condition.wait();
-            }
-        } catch (const ss::condition_variable_timed_out&) {
-            /// There is still more data to truncate
+            co_await _queue.not_empty();
+        } catch (const ss::abort_requested_exception&) {
+            break;
         } catch (const ss::broken_condition_variable&) {
             /// stop() has been called
             break;
+        } catch (const ss::broken_semaphore&) {
```

Comment:
I don't think this is necessary, I think the only exception that could occur from the source within the try block is the one set by  _queue.abort, which is `ss::abort_requested_exception`

Reply:
>I don't think this is necessary, I think the only exception that could occur from the source within the try block is the one set by _queue.abort, which is ss::abort_requested_exception , so we can remove also the catch clause for broken condition variable

agree